src/sundials/sundatanode/sundatanode_inmem.c

```diff
@@ -0,0 +1,475 @@
+  {
+    sunrealtype* data_ptr = leaf_data->ptr;
+    *t                    = data_ptr[0];
+    SUNCheckCall(N_VBufUnpack(v, &data_ptr[1]));
+  }
+  else
+  {
+    SUNMemory buffer_data = NULL;
+    SUNCheckCall(SUNMemoryHelper_Alloc(IMPL_MEMBER(self, mem_helper), &buffer_data,
+                                       buffer_size, buffer_mem_type, queue));
+
+    SUNCheckCall(SUNMemoryHelper_Copy(IMPL_MEMBER(self, mem_helper), buffer_data,
+                                      leaf_data, buffer_size, queue));
+
+    sunrealtype* data_ptr = leaf_data->ptr;
+    *t                    = data_ptr[0];
+    data_ptr              = buffer_data->ptr;
+
+    SUNCheckCall(N_VBufUnpack(v, &data_ptr[1]));
+
+    SUNCheckCall(SUNMemoryHelper_Dealloc(IMPL_MEMBER(self, mem_helper),
+                                         buffer_data, queue));
+  }
+
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNDataNode_SetData_InMem(SUNDataNode self, SUNMemoryType src_mem_type,
+                                     SUNMemoryType node_mem_type, void* data,
+                                     size_t data_stride, size_t data_bytes)
+{
+  SUNFunctionBegin(self->sunctx);
+
+  void* queue = NULL;
+
+  SUNAssert(BASE_MEMBER(self, dtype) == SUNDATANODE_LEAF, SUN_ERR_ARG_WRONGTYPE);
+
+  SUNMemory data_mem_src = SUNMemoryHelper_Wrap(IMPL_MEMBER(self, mem_helper),
+                                                data, src_mem_type);
+  SUNCheckLastErr();
+
+  SUNMemory data_mem_dst = NULL;
+  SUNCheckCall(SUNMemoryHelper_AllocStrided(IMPL_MEMBER(self, mem_helper),
+                                            &data_mem_dst, data_bytes,
+                                            data_stride, node_mem_type, queue));
+
+  SUNCheckCall(SUNMemoryHelper_Copy(IMPL_MEMBER(self, mem_helper), data_mem_dst,
+                                    data_mem_src, data_bytes, queue));
+
+  SUNMemoryHelper_Dealloc(IMPL_MEMBER(self, mem_helper), data_mem_src, queue);
+
+  IMPL_MEMBER(self, leaf_data) = data_mem_dst;
+
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNDataNode_SetDataNvector_InMem(SUNDataNode self, N_Vector v,
+                                            sunrealtype t)
+{
+  SUNFunctionBegin(self->sunctx);
+
+  void* queue = NULL;
+
+  SUNMemoryType leaf_mem_type = SUNMEMTYPE_HOST;
+  SUNMemoryType buffer_mem_type = N_VGetDeviceArrayPointer(v) ? SUNMEMTYPE_DEVICE
+                                                              : SUNMEMTYPE_HOST;
+
+  sunindextype buffer_size = 0;
+  SUNCheckCall(N_VBufSize(v, &buffer_size));
+
+  /* We allocate 1 extra sunrealtype for storing t */
+  SUNMemory leaf_data = NULL;
+  SUNCheckCall(
+    SUNMemoryHelper_AllocStrided(IMPL_MEMBER(self, mem_helper), &leaf_data,
+                                 buffer_size + sizeof(sunrealtype),
+                                 sizeof(sunrealtype), leaf_mem_type, queue));
+
+  if (leaf_mem_type == buffer_mem_type)
+  {
+    sunrealtype* data_ptr = leaf_data->ptr;
+    data_ptr[0]           = t;
```

Comment:
Here and below, can we assume `data_prtr` is host data or does this need to use the memory helper copy?

Reply:
Yes, we can assume so because we define it as host data above.

Reply:
I think I comment in the wrong place, below this is
```
    SUNMemory buffer_data = NULL;
    SUNCheckCall(
      SUNMemoryHelper_AllocStrided(IMPL_MEMBER(self, mem_helper), &buffer_data,
                                   buffer_size + sizeof(sunrealtype),
                                   sizeof(sunrealtype), buffer_mem_type, queue));

    sunrealtype* data_ptr = buffer_data->ptr;
    data_ptr[0]           = t;
    SUNCheckCall(N_VBufPack(v, &data_ptr[1]));
```
If `buffer_mem_type` is `DEVICE` won't setting `data_ptr[0]` segfault?

Reply:
Ah, yes, you are right. We need to set the time in `leaf_data->ptr` (which is on the host), rather than in the `buffer_data->ptr`.